### PR TITLE
fixed issue 3908 by adding trigonometric functions to ParameterExpres…

### LINKED ISSUE #3908: Parameter does not support ufunc à la QASM
### What is the current behavior?

In order to emulate the following QASM, `Parameter` needs to apply along with trigonometry.
```
gate my_gate(phi,lambda) q
            {u1(cos(phi)/2) q;} 
```
A possible way, it would be using `sympy`
```
phi = Parameter('phi') 
circuit.u1(sympy.cos(phi/2), 0)
```
but this raises `CircuitError: 'invalid param type cos in instruction u1'`. Other libraries like `math` and `numpy` require a float-castable parameter. Same applies to the rest of the "QASM ufunc".

### Suggested solutions

Probably, the best solution is to extend `Parameter` type with these ufuncs. So the example would look like:
```
phi = Parameter('phi') 
circuit.u1((phi/2).cos(), 0)
```

### PR DESCRIPTION
**Pull request by georgios-ts**
### Summary

Fixes #3908

### Details and comments

There was no support for trigonometric  functions with a `Parameter` input. In this update, the file with the `ParameterExpression` class includes three new functions: sine, cosine and complex exponential.  The syntax for calling this functions is the same as it was suggested, ie
```python
from qiskit import QuantumCircuit
from qiskit.circuit.parameterexpression import cos

phi = Parameter('phi')
circuit = QuantumCircuit(1)
circuit.u1(cos(phi/2), 0)
```
There was a need (see )  to add  support for multiplication of a `Parameter` with the Imaginary Unit. Consequently, the `conjugate` function was modified accordingly.

### CODE REVIEW

**Comment by Cryoris** [comment]
Hi @georgios-ts, thanks for the contribution!
Could you elaborate a little more on the details of your PR? E.g.
* why are the `sympy.Symbol`s required to be real? 
* is there a reason you put `cos` and `exp` not inside the class `ParameterExpression` but outside? 

Also could you add tests, testing the new methods?

**Comment by georgios-ts** [comment]
> Hi @georgios-ts, thanks for the contribution!
> Could you elaborate a little more on the details of your PR? E.g.
> 
> * why are the `sympy.Symbol`s required to be real?
> * is there a reason you put `cos` and `exp` not inside the class `ParameterExpression` but outside?
> 
> Also could you add tests, testing the new methods?

* I declared `sympy.Symbol`s as real since already in the existing code you can bind only real values to a `Parameter`. The main reason is that in this way the `conjugate` function from `sympy` works nicely and previous tests (that expected  `Parameter` and the `conjugate` of it  to be equal) do not fail. See this example
```python
from sympy import I, Symbol, conjugate
x = Symbol('x')
print(conjugate(x + I))
>> conjugate(x) - I

x = Symbol('x', real=True)
print(conjugate(x + I))
>> x - I
```

* So you can call the function as `cos(phi)` rather than `phi.cos()`.

Yes sure. The `qiskit-terra/test/python/circuit/test_parameters.py` seems to be the right place, correct?

File: qiskit/circuit/parameterexpression.py [hunk 1]
```diff
@@ -262,3 +268,37 @@ def __eq__(self, other):
...
+        raise TypeError('Input should be a ParameterExpression')
+    from sympy import sin as _sin
+
+    return ParameterExpression(
+        self._parameter_symbols,
+        _sin(self._symbol_expr))
+
+
+def exp(self):
+    """Returns a new Expression with sine applied to the old one. """
```
Comment (ewinston) [inline qiskit/circuit/parameterexpression.py:296]: `exp` applied...

File: qiskit/circuit/parameterexpression.py [hunk 2]
```diff
@@ -262,3 +268,37 @@ def __eq__(self, other):
...
+    if not isinstance(self, ParameterExpression):
+        raise TypeError('Input should be a ParameterExpression')
+    from sympy import exp as _exp
+    from sympy import I
+
+    return ParameterExpression(
+        self._parameter_symbols,
+        _exp(I * self._symbol_expr))
```
Comment (ewinston) [inline qiskit/circuit/parameterexpression.py:304]: Probably should not multiply by `I` here for proper exponential.

**Comment by georgios-ts** [comment]
I decided to make a new file for `universal function`s. In this way the code is more clean and can be extended easily. I want also to point out the reason there are distinct functions for `exp` and `complex exp`. If someone call `exp(2*I*phi)` the code will fail since i didn't add support for multiplying `2*I` with `phi`. So it's better for the user to just call `cexp(2*phi)`. The same reasoning follows with similar expressions like `exp(1j * phi)`. Finally, some tests were added.

**Comment by Cryoris** [comment]
With the functions in a separate file users will have to explicitly import the functions to use them. 
However if you make them methods of the `ParameterExpression` class we can also use `numpy.cos`, which is commonly used. I think this gives it a more natural feel.
```python
class ParameterExpression:
    ...
    def cos(self):
        ...

x = Parameter('x')
circuit.ry(np.cos(x))
circuit.rx(np.sin(0.123))
```

**Comment by georgios-ts** [comment]
I merged them back in one file as methods of `ParameterExpression` . Using numpy is really convenient.

I should note that i removed the complex exponentiation since currently, arithmetic operations with complex values (`1j`) are not supported and some tests are checking against it.

**Comment by eggerdj** [comment]
Can we also add support for `max`, `min`, `ceil`, and `floor` while we are at it?

File: qiskit/circuit/parameterexpression.py (hunk 2 above)
Comment (ewinston) [reply qiskit/circuit/parameterexpression.py]: This looks resolved.

**Review by ewinston — APPROVED** [review]

File: qiskit/circuit/parameter.py [hunk 3]
```diff
@@ -45,7 +45,7 @@ def __init__(self, name):
         self._name = name
 
         from sympy import Symbol
-        symbol = Symbol(name)
+        symbol = Symbol(name, real=True)
```
Comment (kdk) [inline qiskit/circuit/parameter.py:48]: Is setting `real=True` required here? This will likely conflict with #5192 .

**Comment by kdk** [comment]
> Can we also add support for `max`, `min`, `ceil`, and `floor` while we are at it?

Can you open a separate issue for these? With complex inputs ( #5192 ) some of these are not as clearly defined, so may require some more thinking.

File: qiskit/circuit/parameter.py (hunk 3 above)
Comment (georgios-ts) [reply qiskit/circuit/parameter.py]: Not required. Resolved.

**Comment by eggerdj** [comment]
> 
> 
> > Can we also add support for `max`, `min`, `ceil`, and `floor` while we are at it?
> 
> Can you open a separate issue for these? With complex inputs ( #5192 ) some of these are not as clearly defined, so may require some more thinking.

@kdk Good point. See here: https://github.com/Qiskit/qiskit-terra/issues/5276

**Review by kdk — APPROVED** [review]

**Pull request merged**